File: demos/52-snapnote/worker/events.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

type eventsConfig struct {
	BaseURL       string
	Consumer      string
	Identity      string
	Subject       string
	AckWaitS      int
	MaxDeliveries int
	PollMS        int
	Batch         int
}

func loadEventsConfig() eventsConfig {
	base := strings.TrimSpace(os.Getenv("FORGE_EVENTS_URL"))
	if base == "" {
		base = "http://host.docker.internal:4105"
	}
	consumer := strings.TrimSpace(os.Getenv("FORGE_EVENTS_CONSUMER"))
	if consumer == "" {
		consumer = "snapnote-attachments"
	}
	identity := strings.TrimSpace(os.Getenv("FORGE_EVENTS_CONSUMER_IDENTITY"))
	if identity == "" {
		identity = consumer
	}
	subject := strings.TrimSpace(os.Getenv("FORGE_EVENTS_SUBJECT"))
	if subject == "" {
		subject = "attachment.uploaded"
	}
	ackWait := envInt("FORGE_DEFAULT_ACK_WAIT_S", 30)
	maxDel := envInt("FORGE_DEFAULT_MAX_DELIVERIES", 5)
	pollMS := envInt("FORGE_EVENTS_POLL_MS", 500)
	batch := envInt("FORGE_EVENTS_BATCH", 8)
	return eventsConfig{
		BaseURL:       strings.TrimRight(base, "/"),
		Consumer:      consumer,
		Identity:      identity,
		Subject:       subject,
		AckWaitS:      ackWait,
		MaxDeliveries: maxDel,
		PollMS:        pollMS,
		Batch:         batch,
	}
}

func envInt(key string, def int) int {
	raw := strings.TrimSpace(os.Getenv(key))
	if raw == "" {
		return def
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n <= 0 {
		return def
	}
	return n
}

type eventsClient struct {
	cfg        eventsConfig
	httpClient *http.Client
}

func newEventsClient(cfg eventsConfig) *eventsClient {
	return &eventsClient{
		cfg: cfg,
		httpClient: &http.Client{
			Timeout: 45 * time.Second,
		},
	}
}

func (c *eventsClient) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.cfg.BaseURL+"/health/ready", nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("events ready HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

func (c *eventsClient) EnsureConsumer(ctx context.Context) error {
	body := map[string]any{
		"name":           c.cfg.Consumer,
		"subject":        c.cfg.Subject,
		"ack_wait_s":     c.cfg.AckWaitS,
		"max_deliveries": c.cfg.MaxDeliveries,
		"identity":       c.cfg.Identity,
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/consumers", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusCreated || resp.StatusCode == http.StatusOK {
		return nil
	}
	b, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
	return fmt.Errorf("create consumer HTTP %d: %s", resp.StatusCode, string(b))
}

type deliveredMessage struct {
	EventID       string          `json:"event_id"`
	Subject       string          `json:"subject"`
	AckToken      string          `json:"ack_token"`
	DeliveryCount int             `json:"delivery_count"`
	Data          json.RawMessage `json:"data"`
}

type consumeResponse struct {
	Messages []deliveredMessage `json:"messages"`
}

func (c *eventsClient) Consume(ctx context.Context) ([]deliveredMessage, error) {
	body := map[string]any{
		"consumer": c.cfg.Consumer,
		"batch":    c.cfg.Batch,
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/consume", bytes.NewReader(raw))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, err := io.ReadAll(io.LimitReader(resp.Body, 4<<20))
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("consume HTTP %d: %s", resp.StatusCode, string(respBody))
	}
	var out consumeResponse
	if err := json.Unmarshal(respBody, &out); err != nil {
		return nil, err
	}
	return out.Messages, nil
}

func (c *eventsClient) MarkProcessed(ctx context.Context, eventID string) error {
	body := map[string]any{
		"consumer": c.cfg.Consumer,
		"event_id": eventID,
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/processed", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("processed HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

func (c *eventsClient) Ack(ctx context.Context, ackToken string) error {
	body := map[string]any{"ack_token": ackToken}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/ack", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("ack HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

func (c *eventsClient) Nak(ctx context.Context, ackToken string, delayS int) error {
	body := map[string]any{"ack_token": ackToken}
	if delayS > 0 {
		body["delay_s"] = delayS
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/nak", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("nak HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

File: demos/53-askdocs/worker/events.py
"""Forge Events consumer client for AskDocs ingest worker (epic 53.02)."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any


@dataclass
class EventsConfig:
    base_url: str
    consumer: str
    identity: str
    subject: str
    ack_wait_s: int
    max_deliveries: int
    poll_ms: int
    batch: int


def _env_int(env: dict[str, str], key: str, default: int) -> int:
    raw = (env.get(key) or "").strip()
    if not raw:
        return default
    try:
        n = int(raw)
    except ValueError:
        return default
    return n if n > 0 else default


def load_events_config(environ: dict[str, str] | None = None) -> EventsConfig:
    env = environ if environ is not None else os.environ
    base = (env.get("FORGE_EVENTS_URL") or "").strip() or "http://host.docker.internal:4105"
    consumer = (env.get("FORGE_EVENTS_CONSUMER") or "").strip() or "askdocs-documents"
    identity = (env.get("FORGE_EVENTS_CONSUMER_IDENTITY") or "").strip() or consumer
    subject = (env.get("FORGE_EVENTS_SUBJECT") or "").strip() or "document.uploaded"
    return EventsConfig(
        base_url=base.rstrip("/"),
        consumer=consumer,
        identity=identity,
        subject=subject,
        ack_wait_s=_env_int(env, "FORGE_DEFAULT_ACK_WAIT_S", 30),
        max_deliveries=_env_int(env, "FORGE_DEFAULT_MAX_DELIVERIES", 5),
        poll_ms=_env_int(env, "FORGE_EVENTS_POLL_MS", 500),
        batch=_env_int(env, "FORGE_EVENTS_BATCH", 8),
    )


@dataclass
class DeliveredMessage:
    event_id: str
    subject: str
    ack_token: str
    delivery_count: int
    data: dict[str, Any]


class EventsClient:
    def __init__(self, cfg: EventsConfig | None = None) -> None:
        self.cfg = cfg or load_events_config()

    def _request(
        self,
        method: str,
        url: str,
        data: bytes | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 45.0,
    ) -> tuple[int, bytes]:
        req = urllib.request.Request(url, data=data, method=method, headers=headers or {})
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return int(resp.status), resp.read()
        except urllib.error.HTTPError as exc:
            body = exc.read() if exc.fp else b""
            return int(exc.code), body

    def ping(self) -> None:
        code, body = self._request("GET", f"{self.cfg.base_url}/health/ready", timeout=10.0)
        if code != 200:
            raise RuntimeError(f"events ready HTTP {code}: {body[:256]!r}")

    def ensure_consumer(self) -> None:
        payload = {
            "name": self.cfg.consumer,
            "subject": self.cfg.subject,
            "ack_wait_s": self.cfg.ack_wait_s,
            "max_deliveries": self.cfg.max_deliveries,
            "identity": self.cfg.identity,
        }
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/consumers",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 201):
            raise RuntimeError(f"create consumer HTTP {code}: {body[:512]!r}")

    def consume(self) -> list[DeliveredMessage]:
        payload = {"consumer": self.cfg.consumer, "batch": self.cfg.batch}
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/consume",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=45.0,
        )
        if code != 200:
            raise RuntimeError(f"consume HTTP {code}: {body[:512]!r}")
        parsed = json.loads(body.decode() or "{}")
        out: list[DeliveredMessage] = []
        for item in parsed.get("messages") or []:
            data = item.get("data") or {}
            if isinstance(data, str):
                try:
                    data = json.loads(data)
                except json.JSONDecodeError:
                    data = {}
            if not isinstance(data, dict):
                data = {}
            out.append(
                DeliveredMessage(
                    event_id=str(item.get("event_id") or ""),
                    subject=str(item.get("subject") or ""),
                    ack_token=str(item.get("ack_token") or ""),
                    delivery_count=int(item.get("delivery_count") or 0),
                    data=data,
                )
            )
        return out

    def mark_processed(self, event_id: str) -> None:
        raw = json.dumps({"consumer": self.cfg.consumer, "event_id": event_id}).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/processed",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"processed HTTP {code}: {body[:256]!r}")

    def ack(self, ack_token: str) -> None:
        raw = json.dumps({"ack_token": ack_token}).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/ack",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"ack HTTP {code}: {body[:256]!r}")

    def nak(self, ack_token: str, delay_s: int = 0) -> None:
        payload: dict[str, Any] = {"ack_token": ack_token}
        if delay_s > 0:
            payload["delay_s"] = delay_s
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/nak",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"nak HTTP {code}: {body[:256]!r}")
